src/query.rs

```diff
@@ -762,6 +828,195 @@ macro_rules! tuple_impl {
 //smaller_tuples_too!(tuple_impl, B, A);
 smaller_tuples_too!(tuple_impl, O, N, M, L, K, J, I, H, G, F, E, D, C, B, A);
 
+/// A prepared query caches some of the information required to execute the query `Q`
+pub struct PreparedQuery<Q: Query> {
+    memo: (usize, u64),
+    state: Box<[(usize, <Q::Fetch as Fetch<'static>>::State)]>,
+    _query: PhantomData<Q>,
+}
+
+impl<Q: Query> PreparedQuery<Q> {
+    #[cold]
+    fn new(world: &World) -> Self {
+        let memo = world.memo();
+
+        let state = world
+            .archetypes()
+            .enumerate()
+            .filter_map(|(idx, x)| Q::Fetch::prepare(x).map(|state| (idx, state)))
+            .collect();
+
+        Self {
+            memo,
+            state,
+            _query: PhantomData,
+        }
+    }
+
+    /// Borrow the query and the [world][World] as a prerequisite for execution
+    ///
+    /// This will panic if it would violate an existing unique reference
+    /// or construct an invalid unique reference.
+    pub fn borrow<'q>(&'q mut self, world: &'q World) -> PreparedQueryBorrow<'q, Q> {
```

Comment:
Let's use terminology consistent with non-prepared queries:
```suggestion
    /// Query `world`, using dynamic borrow checking
    ///
    /// This will panic if it would violate an existing unique reference
    /// or construct an invalid unique reference.
    pub fn query<'q>(&'q mut self, world: &'q World) -> PreparedQueryBorrow<'q, Q> {
```